eevee/metrics.py

```diff
@@ -1,20 +1,59 @@
-from typing import List
+from typing import Callable, List, Optional
 
+import numpy as np
 from sklearn.metrics import confusion_matrix
 
 from eevee.types import SlotLabel
 
 
-def slot_capture_rate() -> float:
-    ...
+def slot_capture_rate(slots_predicted: List[SlotLabel], expected_slot_value: str) -> float:
+    """
+    """
 
+    if isinstance(slots_predicted, list) and slots_predicted:
+        captured_count = slots_predicted.count(expected_slot_value)
+        return captured_count / len(slots_predicted)
+    else:
+        raise ValueError(f"Expected {slots_predicted = } to be a list of size greater than 0")
 
-def slot_retry_rate() -> float:
-    ...
+
+def slot_retry_rate(slot_turn_counts: List[Optional[int]], agg_fn: Callable = np.mean) -> float:
+    """
+    Slot retry rate refers to aggregating on the number of turns a slot requires.
+
+    Let `n` be the number of calls,
+    `slot_turn_counts` be number of turns required to handle the particular slot in each call.
+    `agg_fn` runs aggregation on `slot_turn_counts` and returns a float.
+    """
+
+    if isinstance(slot_turn_counts, list) and slot_turn_counts:
+        filtered_turn_counts = list(filter(None, slot_turn_counts))
+        return agg_fn(filtered_turn_counts)
+    else:
+        raise ValueError(f"Expected {slot_turn_counts =} to be a list of size greater than 0")
 
 
 def slot_mismatch_rate(y_true: List[SlotLabel], y_pred: List[SlotLabel]) -> float:
-    ...
+    """
+    As per the definition, slot mismatch rate captures
+    ratio of match on types-but-not-values with
+    match on types-but-not-values + match on types-and-values
+    """
+
+    type_and_value_matched = 0
+    type_matched_but_value_didnt = 0
+
+    for y_true_i, y_pred_i in zip(y_true, y_pred):
+
+        if type(y_true_i) == type(y_pred_i) and y_true_i is not None:
```

Comment:
This will need update. Type is based on the slot's entity type. Since that's not yet defined, we will let it pass in the PR.

Reply:
We would actually need equality functions for various entity types. Check [here](https://gitlab.com/vernacularai/ai/plute/-/tree/master/plute/ord/) for examples.